lib/pacer_test.go

```diff
@@ -0,0 +1,148 @@
+		}
+	}
+}
+
+var quarterPeriods = map[string]float64{
+	"MeanUp":   MeanUp,
+	"Peak":     Peak,
+	"MeanDown": MeanDown,
+	"Trough":   Trough,
+}
+
+type sineTest struct {
+	p, m, a int
+}
+
+func (st sineTest) Pacer(startAt float64) SinePacer {
+	return SinePacer{
+		Period:  time.Duration(st.p) * time.Second,
+		Mean:    Rate{st.m, time.Second},
+		Amp:     Rate{st.a, time.Second},
+		StartAt: startAt,
+	}
+}
+
+func (st sineTest) AmpHits() float64 {
+	return float64(st.a*st.p) / (2 * math.Pi)
+}
+
+func (st sineTest) Hits(frac, startAt float64) uint64 {
+	return uint64(math.Round(
+		float64(st.m*st.p)*frac +
+			st.AmpHits()*(math.Cos(startAt)-math.Cos(startAt+frac*2*math.Pi))))
+}
+
+func (st sineTest) Nanos(frac, startAt float64) time.Duration {
+	return time.Duration(1 / (float64(st.m) + float64(st.a)*math.Sin(startAt+frac*2*math.Pi)))
+}
+
+func TestSinePacerHits(t *testing.T) {
+	tests := []sineTest{
+		{20 * 60, 100, 90},
+		{60, 1000, 10},
+		{1, 10, 7},
+		{1, 1, 0},
+		// These test cases failed with off-by-one errors before applying
+		// math.Round in Hits, due to floating-point maths differences.
+		{1e6, 10, 7},
+		{60, 1000, 999},
+	}
+
+	for i, test := range tests {
+		for name, sa := range quarterPeriods {
+			sr := test.Pacer(sa)
+			if got, want := sr.Hits(sr.Period/4), test.Hits(0.25, sa); got != want {
+				t.Errorf("%d(%s): hits after 1/4 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(sr.Period/2), test.Hits(0.5, sa); got != want {
+				t.Errorf("%d(%s): hits after 1/2 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(3*sr.Period/4), test.Hits(0.75, sa); got != want {
+				t.Errorf("%d(%s): hits after 3/4 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(sr.Period), test.Hits(1, sa); got != want {
+				t.Errorf("%d(%s): hits after full period = %d, want %d", i, name, got, want)
+			}
+		}
+	}
+}
+
+func TestSinePacerFlat(t *testing.T) {
+	st := sineTest{1, 1, 0}
+	tests := []struct {
+		et   time.Duration
+		c    uint64
+		wait time.Duration
+		stop bool
+	}{
+		{0, 0, time.Second, false},
+		{0, 1, 2 * time.Second, false},
+		{time.Second / 100, 0, 99 * time.Second / 100, false},
+		{time.Second / 2, 0, time.Second / 2, false},
+		{64 * time.Second / 100, 0, 36 * time.Second / 100, false},
+		// Has an off-by-one I can't round away nicely because it's
+		// due to expectedHits being 0.9900000000000001. Ugh floats.
+		// {99 * time.Second / 100, 0, time.Second / 100},
+		{time.Second, 1, time.Second, false},
+		{time.Second, 0, 0, false},
+	}
+
+	for i, test := range tests {
+		for name, sa := range quarterPeriods {
+			p := st.Pacer(sa)
+			wait, stop := p.Pace(test.et, test.c)
+			if wait != test.wait || stop != test.stop {
+				t.Errorf("%d(%s): wait(%v) = (%v, %v), want (%v, %v)",
+					i, name, test.et, wait, stop, test.wait, test.stop)
+			}
+		}
+	}
+}
+
+// No idea how to test interval without hard-coding a bunch of numbers.
```

Comment:
Neither would I :D

Reply:
Property-based testing could help us out here (look up some presentations on youtube).

The basic idea is:
- instead of using concrete examples to test our code we figure out some properties for our code that absolutely must hold, and we describe these properties in tests.

Here are some examples ... suppose I have time interval of a certain predefined `length`, I also have fixed rate pacer (with the rate, say, `freq`/`per`):
- then I know that for any `freq` and `per` (we can generate them randomly) and any `lenght` if I call `Pace(length, x)` where `x = any from 0 to (freq * (length / per) - 1)` I should get `(0, false)` back, meaning pacer has fallen behind and must return no "wait time" at all;
- then I know that for any `freq` and `per` (we can generate them randomly) and any `lenght` if I call `Pace(length, x)` where `x = any from (freq * (length / per) + 1) to MAX_INT` I should get `(not zero, false)` back, meaning pacer is ahead and has no right to return 0 for "wait time"
- then I know that while I'm pacing some interval I cannot accumulate "wait time" greater than the length for that interval
- etc.

Reply:
There probably exist some testing frameworks for property-based testing, but here is fast&dirty way to test the properties I described above:

```go
func TestFixedRatePacerPropertyBased(t *testing.T) {
	rand.Seed(time.Now().UnixNano())

	t.Run("when pacer has fallen behind it must return zero wait time", func(t *testing.T) {
		for i := 0; i < 100; i++ {
			freq := 1 + rand.Intn(math.MaxInt32-1)
			per := time.Duration(1 + rand.Int63n(math.MaxInt64-1))
			// time interval length
			length := time.Duration((int64(per)/int64(freq) + 1) + rand.Int63n(10000))

			expectedHitsForInterval := float64(freq) * (float64(length) / float64(per))
			hitsDone := uint64(rand.Int63n(int64(math.Floor(expectedHitsForInterval))))

			r := &Rate{Freq: freq, Per: per}
			wait, stop := r.Pace(length, hitsDone)
			if stop == true {
				t.Errorf("freq=%v per=%v length=%v, stop: exp = %v, act = %v", freq, per, length, false, stop)
			}
			if wait > 0 {
				t.Errorf("freq=%v per=%v length=%v, wait: exp = zero or negative number, act = %v", freq, per, length, wait)
			}
		}
	})
	t.Run("when pacer is ahead it must return positive wait time", func(t *testing.T) {
		for i := 0; i < 100; i++ {
			freq := 1 + rand.Intn(math.MaxInt32-1)
			per := time.Duration(1 + rand.Int63n(math.MaxInt64-1))
			// time interval length
			length := time.Duration((int64(per)/int64(freq) + 1) + rand.Int63n(10000))

			expectedHitsForInterval := float64(freq) * (float64(length) / float64(per))
			boundary := int64(math.Ceil(expectedHitsForInterval))
			hitsDone := uint64(boundary + rand.Int63n(math.MaxInt64-boundary))
			//hitsDone := uint64(boundary + 22222222)

			r := &Rate{Freq: freq, Per: per}
			wait, stop := r.Pace(length, hitsDone)
			if stop == true {
				t.Errorf("freq=%v per=%v length=%v, stop: exp = %v, act = %v", freq, per, length, false, stop)
			}
			if wait <= 0 {
				t.Errorf("freq=%v per=%v length=%v, wait: exp = positive number, act = %v", freq, per, length, wait)
			}
		}
	})
	t.Run("pacing some interval I cannot accumulate wait time greater than the length for that interval", func(t *testing.T) {
		for i := 0; i < 100; i++ {
			freq := 1 + rand.Intn(math.MaxInt32-1)
			per := time.Duration(1 + rand.Int63n(math.MaxInt64-1))
			// time interval length
			length := time.Duration((int64(per)/int64(freq) + 1) + rand.Int63n(10000))

			elapsed := time.Duration(0)
			hits := uint64(0)

			maxHitsExp := uint64(math.Ceil((float64(length) / float64(per)) * float64(freq)))

			r := &Rate{Freq: freq, Per: per}
			for hits < maxHitsExp {
				wait, stop := r.Pace(length, hits)
				if stop == true {
					t.Errorf("freq=%v per=%v length=%v, stop: exp = %v, act = %v", freq, per, length, false, stop)
				}
				elapsed += wait
				hits++
			}

			if elapsed > length {
				t.Errorf("freq=%v per=%v length=%v, wait: elapsed = %v", freq, per, length, elapsed)
			}
		}
	})
}
```

Some of them fails probably because of overflows (haven't had time to look). Also I'd probably reduce some code duplication in these tests if they get through to the final PR.

Reply:
If you like this approach, I coult tidy things up and write some more property-based tests.

Reply:
Interesting! Does it extend nicely to variable rates? The first two look like they should as long as you can determine expectedHits from elapsedTime for your Pacer, but I can't tell as easily for the third one.

I'd definitely be grateful for a cleaned-up test suite that you could just feed pacer implementations to.

Reply:
> but I can't tell as easily for the third one

I think it does ... we can run it several times and be pretty sure whether or not :)

> Does it extend nicely to variable rates?

Some properties will be true for all `Pacer` implementations, others may be specific to only few, it depends on what properties we will come up with to validate our code :)

The idea of creating properties that every `Pacer` must satisfy feels good to me as we will be able to test new `Pacer` implementation just by inserting them in our test-suite.

If you any other properties (other that 3 listed above) come to your mind - let me know :)

Reply:
> If you like this approach, I could tidy things up and write some more property-based tests.

> The idea of creating properties that every Pacer must satisfy feels good to me as we will be able to test new Pacer implementation just by inserting them in our test-suite.

Yes and YES! Property based testing is awesome. I use it a lot in another OSS project I maintain. Example: https://github.com/oklog/ulid/blob/master/ulid_test.go#L116

This level of contribution goes beyond my wildest dreams. Thank you both!